core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -930,30 +941,46 @@ public PersistenceState getPersistenceState(long fileId) throws FileDoesNotExist
       auditContext.setSrcInode(inode);
 
       List<FileInfo> ret = new ArrayList<>();
-      if (inode.isDirectory()) {
-        TempInodePathForDescendant tempInodePath = new TempInodePathForDescendant(inodePath);
-        try {
-          mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, inodePath);
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
-          throw e;
-        }
-        for (Inode<?> child : ((InodeDirectory) inode).getChildren()) {
-          child.lockReadAndCheckParent(inode);
+      listStatusRecursive(ret, inodePath, auditContext, listStatusOptions.isRecursive());
+
+      auditContext.setSucceeded(true);
+      Metrics.FILE_INFOS_GOT.inc();
+      return ret;
+    }
+  }
+
+  private void listStatusRecursive(List<FileInfo> statusList, LockedInodePath originalInodePath,
+                                   AuditContext auditContext, boolean recursive)
+      throws UnavailableException, FileDoesNotExistException,
+      AccessControlException, InvalidPathException {
+    Queue<LockedInodePath> queue = new LinkedList<>();
+    queue.add(originalInodePath);
+
+    try (InodeLockList lockList = new InodeLockList()) {
+      while (!queue.isEmpty()) {
+        LockedInodePath currInodePath = queue.remove();
+        Inode<?> inode = currInodePath.getInode();
+        if (inode.isDirectory()) {
           try {
-            // the path to child for getPath should already be locked.
-            tempInodePath.setDescendant(child, mInodeTree.getPath(child));
-            ret.add(getFileInfoInternal(tempInodePath));
-          } finally {
-            child.unlockRead();
+            mPermissionChecker.checkPermission(Mode.Bits.EXECUTE, originalInodePath);
+          } catch (Exception e) {
+            auditContext.setAllowed(false);
```

Comment:
How does this work on a local FS? If you do a `ls -R`, and you don't access to some dirs, does the ls fail, or do you see the ones you do have access to?

Reply:
I tried it on a local FS. it fails for some and returns the result for others that you do have access to.  I can return result for those ones that you do have access to, but will silently fail for those that you don't have access to.